contracts/RolesRegistry/SftRolesRegistry.sol

```diff
@@ -0,0 +1,237 @@
+            require(deposits[_grantRoleData.nonce].tokenId == _grantRoleData.tokenId, "SftRolesRegistry: tokenId mismatch");
+            require(deposits[_grantRoleData.nonce].tokenAmount == _grantRoleData.tokenAmount, "SftRolesRegistry: tokenAmount mismatch");
+
+            RoleData storage _roleData = roleAssignments[_grantRoleData.nonce];
+            require(_roleData.expirationDate < block.timestamp || _roleData.revocable, "SftRolesRegistry: nonce is not expired or is not revocable");
+        }
+        _grantOrUpdateRole(_grantRoleData);
+    }
+
+    function _grantOrUpdateRole(RoleAssignment calldata _grantRoleData) internal {
+        roleAssignments[_grantRoleData.nonce] = RoleData(
+            _grantRoleData.grantee,
+            _grantRoleData.expirationDate,
+            _grantRoleData.revocable,
+            _grantRoleData.data
+        );
+
+        emit RoleGranted(
+            _grantRoleData.nonce,
+            UNIQUE_ROLE,
+            _grantRoleData.tokenAddress,
+            _grantRoleData.tokenId,
+            _grantRoleData.tokenAmount,
+            _grantRoleData.grantor,
+            _grantRoleData.grantee,
+            _grantRoleData.expirationDate,
+            _grantRoleData.revocable,
+            _grantRoleData.data
+        );
+    }
+
+    function _deposit(uint256 _nonce, DepositInfo memory _depositInfo) internal {
+        deposits[_nonce] = _depositInfo;
+
+        _transferFrom(
+            _depositInfo.grantor,
+            address(this),
+            _depositInfo.tokenAddress,
+            _depositInfo.tokenId,
+            _depositInfo.tokenAmount
+        );
+    }
+
+    function revokeRoleFrom(uint256 _nonce, bytes32 _role, address _grantee) external override {
+        RoleData memory _roleData = roleAssignments[_nonce];
+        require(_roleData.grantee != address(0), 'SftRolesRegistry: invalid grantee');
+        DepositInfo memory _depositInfo = deposits[_nonce];
+
+        address caller = _findCaller(_roleData, _depositInfo);
+        if (_roleData.expirationDate > block.timestamp && !_roleData.revocable) {
+            // if role is not expired and is not revocable, only the grantee can revoke it
+            require(caller == _roleData.grantee, 'SftRolesRegistry: nonce is not expired or is not revocable');
+        }
+
+        delete roleAssignments[_nonce];
+
+        emit RoleRevoked(
+            _nonce,
+            UNIQUE_ROLE,
+            _depositInfo.tokenAddress,
+            _depositInfo.tokenId,
+            _depositInfo.tokenAmount,
+            _depositInfo.grantor,
+            _roleData.grantee
+        );
+    }
+
+    function withdraw(uint256 _nonce)
+        public
+        onlyOwnerOrApprovedWithBalance(
+            deposits[_nonce].grantor,
+            deposits[_nonce].tokenAddress,
+            deposits[_nonce].tokenId,
+            deposits[_nonce].tokenAmount
+        )
```

Comment:
We need to check if caller is approved to make this call or not. However the `onlyOwnerOrApprovedWithBalance` modifier goes beyond that and also checks the `tokenAmount`